File: tst/tile-grid.c
#include "pico.h"
#include "../check.h"

int main (void) {
    pico_init(1);

    pico_set_window((Pico_Window){ .dim={480,480}, .fs=0, .show=1, .title="Grid" });

    // Test 1: main layer, no tile mode, pixel grid only
    {
        puts("pixel grid only (no tile)");
        Pico_Rel_Dim log = { '!', {16, 16} };
        pico_set_effect_grid(1); pico_set_scene_dim(&log);
        pico_output_clear();
        pico_set_pencil_color(PICO_COLOR_RED);
        pico_output_draw_rect(&(Pico_Rel_Rect){ '!', {4, 4, 8, 8}, PICO_ANCHOR_NW });
        _pico_check("tile-grid-01");
    }

    // Test 2: main layer, 4x4 tiles of 4x4 pixels, pixel + tile grid
    {
        puts("pixel + tile grid (4x4 tiles)");
        Pico_Rel_Dim log = { '#', {4, 4} };
        Pico_Abs_Dim tile = { 4, 4 };
        pico_set_effect_grid(1); pico_set_scene_tile(tile); pico_set_scene_dim(&log);
        pico_output_clear();
        pico_set_pencil_color(PICO_COLOR_GREEN);
        pico_output_draw_rect(&(Pico_Rel_Rect){ '#', {1, 1, 2, 2}, PICO_ANCHOR_NW });
        _pico_check("tile-grid-02");
    }

    // Test 3: main layer, 8x8 tiles of 2x2 pixels, pixel + tile grid
    {
        puts("pixel + tile grid (8x8 tiles)");
        Pico_Rel_Dim log = { '#', {8, 8} };
        Pico_Abs_Dim tile = { 2, 2 };
        pico_set_effect_grid(1); pico_set_scene_tile(tile); pico_set_scene_dim(&log);
        pico_output_clear();
        pico_set_pencil_color(PICO_COLOR_BLUE);
        pico_output_draw_rect(&(Pico_Rel_Rect){ '#', {3, 3, 2, 2}, PICO_ANCHOR_NW });
        _pico_check("tile-grid-03");
    }

    // Test 4: main layer, non-square tiles (4x2), pixel + tile grid
    {
        puts("pixel + tile grid (non-square 4x2 tiles)");
        Pico_Rel_Dim log = { '#', {4, 8} };
        Pico_Abs_Dim tile = { 4, 2 };
        pico_set_effect_grid(1); pico_set_scene_tile(tile); pico_set_scene_dim(&log);
        pico_output_clear();
        pico_set_pencil_color(PICO_COLOR_YELLOW);
        pico_output_draw_rect(&(Pico_Rel_Rect){ '#', {2, 4, 1, 2}, PICO_ANCHOR_NW });
        _pico_check("tile-grid-04");
    }

    // Test 5: main layer, grid disabled, no grid lines
    {
        puts("grid disabled");
        Pico_Rel_Dim log = { '#', {4, 4} };
        Pico_Abs_Dim tile = { 4, 4 };
        pico_set_effect_grid(0); pico_set_scene_tile(tile); pico_set_scene_dim(&log);
        pico_output_clear();
        pico_set_pencil_color(PICO_COLOR_RED);
        pico_output_draw_rect(&(Pico_Rel_Rect){ '#', {2, 2, 1, 1}, PICO_ANCHOR_C });
        _pico_check("tile-grid-05");
    }

    // Reset main to simple view for layer tests, grid off
    {
        Pico_Rel_Dim log = { '!', {480, 480} };
        pico_set_effect_grid(0); pico_set_scene_dim(&log);
    }

    // Layer A: 4x4 tiles of 4x4 pixels, grid enabled
    {
        puts("layer A: 4x4 tiles, grid on");
        pico_layer_empty(NULL, "layerA", (Pico_Abs_Dim){16, 16}, NULL);
        const char* old = pico_set_layer("layerA");
        Pico_Rel_Dim log = { '#', {4, 4} };
        Pico_Abs_Dim tile = { 4, 4 };
        pico_set_effect_grid(1); pico_set_scene_tile(tile); pico_set_scene_dim(&log);
        pico_output_clear();
        pico_set_pencil_color(PICO_COLOR_RED);
        pico_output_draw_rect(&(Pico_Rel_Rect){ '#', {1, 1, 2, 2}, PICO_ANCHOR_NW });
        pico_set_layer(old);
    }

    // Layer B: 8x4 tiles of 2x4 pixels, grid enabled
    {
        puts("layer B: 8x4 non-square tiles, grid on");
        pico_layer_empty(NULL, "layerB", (Pico_Abs_Dim){16, 16}, NULL);
        const char* old = pico_set_layer("layerB");
        Pico_Rel_Dim log = { '#', {8, 4} };
        Pico_Abs_Dim tile = { 2, 4 };
        pico_set_effect_grid(1); pico_set_scene_tile(tile); pico_set_scene_dim(&log);
        pico_output_clear();
        pico_set_pencil_color(PICO_COLOR_GREEN);
        pico_output_draw_rect(&(Pico_Rel_Rect){ '#', {5, 2, 2, 1}, PICO_ANCHOR_NW });
        pico_set_layer(old);
    }

    // Layer C: 4x4 tiles of 4x4 pixels, grid disabled
    {
        puts("layer C: 4x4 tiles, grid off");
        pico_layer_empty(NULL, "layerC", (Pico_Abs_Dim){16, 16}, NULL);
        const char* old = pico_set_layer("layerC");
        Pico_Rel_Dim log = { '#', {4, 4} };
        Pico_Abs_Dim tile = { 4, 4 };
        pico_set_effect_grid(0); pico_set_scene_tile(tile); pico_set_scene_dim(&log);
        pico_output_clear();
        pico_set_pencil_color(PICO_COLOR_BLUE);
        pico_output_draw_rect(&(Pico_Rel_Rect){ '#', {3, 3, 1, 1}, PICO_ANCHOR_NW });
        pico_set_layer(old);
    }

    // Test 6: draw all 3 layers side by side on main
    {
        puts("3 layers: A(grid on) B(grid on) C(grid off)");
        pico_output_clear();
        pico_output_draw_layer("layerA", &(Pico_Rel_Rect){ '!', {0,   0, 150, 150}, PICO_ANCHOR_NW });
        pico_output_draw_layer("layerB", &(Pico_Rel_Rect){ '!', {160, 0, 150, 150}, PICO_ANCHOR_NW });
        pico_output_draw_layer("layerC", &(Pico_Rel_Rect){ '!', {320, 0, 150, 150}, PICO_ANCHOR_NW });
        _pico_check("tile-grid-06");
    }

    // Test 7: layers at different scales
    {
        puts("3 layers: different scales");
        pico_output_clear();
        pico_output_draw_layer("layerA", &(Pico_Rel_Rect){ '!', {0,   170, 120, 120}, PICO_ANCHOR_NW });
        pico_output_draw_layer("layerB", &(Pico_Rel_Rect){ '!', {140, 170, 200, 200}, PICO_ANCHOR_NW });
        pico_output_draw_layer("layerC", &(Pico_Rel_Rect){ '!', {0,   310, 240, 120}, PICO_ANCHOR_NW });
        _pico_check("tile-grid-07");
    }

    pico_init(0);
    return 0;
}

File: lua/tst/tile-grid.lua
require 'pico.check'

pico.init(true)

local phy = {'!', w=480, h=480}
pico.set.window { title="Grid", dim=phy }

-- Test 1: main layer, no tile mode, pixel grid only
do
    print("pixel grid only (no tile)")
    pico.set.show { grid=true }
    pico.set.view { dim={'!', w=16, h=16} }
    pico.output.clear()
    pico.set.draw { color='red' }
    pico.output.draw.rect {'!', x=4, y=4, w=8, h=8, anchor='NW'}
    pico.check("tile-grid-01")
end

-- Test 2: main layer, 4x4 tiles of 4x4 pixels, pixel + tile grid
do
    print("pixel + tile grid (4x4 tiles)")
    pico.set.show { grid=true }
    pico.set.view { dim={'#', w=4, h=4}, tile={w=4, h=4} }
    pico.output.clear()
    pico.set.draw { color='green' }
    pico.output.draw.rect {'#', x=1, y=1, w=2, h=2, anchor='NW'}
    pico.check("tile-grid-02")
end

-- Test 3: main layer, 8x8 tiles of 2x2 pixels, pixel + tile grid
do
    print("pixel + tile grid (8x8 tiles)")
    pico.set.show { grid=true }
    pico.set.view { dim={'#', w=8, h=8}, tile={w=2, h=2} }
    pico.output.clear()
    pico.set.draw { color='blue' }
    pico.output.draw.rect {'#', x=3, y=3, w=2, h=2, anchor='NW'}
    pico.check("tile-grid-03")
end

-- Test 4: main layer, non-square tiles (4x2), pixel + tile grid
do
    print("pixel + tile grid (non-square 4x2 tiles)")
    pico.set.show { grid=true }
    pico.set.view { dim={'#', w=4, h=8}, tile={w=4, h=2} }
    pico.output.clear()
    pico.set.draw { color='yellow' }
    pico.output.draw.rect {'#', x=2, y=4, w=1, h=2, anchor='NW'}
    pico.check("tile-grid-04")
end

-- Test 5: main layer, grid disabled, no grid lines
do
    print("grid disabled")
    pico.set.show { grid=false }
    pico.set.view { dim={'#', w=4, h=4}, tile={w=4, h=4} }
    pico.output.clear()
    pico.set.draw { color='red' }
    pico.output.draw.rect {'#', x=2, y=2, w=1, h=1, anchor='C'}
    pico.check("tile-grid-05")
end

-- Reset main to simple view for layer tests, grid off
pico.set.show { grid=false }
    pico.set.view { dim={'!', w=480, h=480} }

-- Layer A: 4x4 tiles of 4x4 pixels, grid enabled
do
    print("layer A: 4x4 tiles, grid on")
    pico.layer.empty(nil, "layerA", {w=16, h=16})
    pico.set.layer("layerA")
    pico.set.show { grid=true }
    pico.set.view { dim={'#', w=4, h=4}, tile={w=4, h=4} }
    pico.output.clear()
    pico.set.draw { color='red' }
    pico.output.draw.rect {'#', x=1, y=1, w=2, h=2, anchor='NW'}
    pico.set.layer(nil)
end

-- Layer B: 8x4 non-square tiles, grid enabled
do
    print("layer B: 8x4 non-square tiles, grid on")
    pico.layer.empty(nil, "layerB", {w=16, h=16})
    pico.set.layer("layerB")
    pico.set.show { grid=true }
    pico.set.view { dim={'#', w=8, h=4}, tile={w=2, h=4} }
    pico.output.clear()
    pico.set.draw { color='green' }
    pico.output.draw.rect {'#', x=5, y=2, w=2, h=1, anchor='NW'}
    pico.set.layer(nil)
end

-- Layer C: 4x4 tiles of 4x4 pixels, grid disabled
do
    print("layer C: 4x4 tiles, grid off")
    pico.layer.empty(nil, "layerC", {w=16, h=16})
    pico.set.layer("layerC")
    pico.set.show { grid=false }
    pico.set.view { dim={'#', w=4, h=4}, tile={w=4, h=4} }
    pico.output.clear()
    pico.set.draw { color='blue' }
    pico.output.draw.rect {'#', x=3, y=3, w=1, h=1, anchor='NW'}
    pico.set.layer(nil)
end

-- Test 6: draw all 3 layers side by side on main
do
    print("3 layers: A(grid on) B(grid on) C(grid off)")
    pico.output.clear()
    pico.output.draw.layer("layerA", {'!', x=0,   y=0, w=150, h=150, anchor='NW'})
    pico.output.draw.layer("layerB", {'!', x=160, y=0, w=150, h=150, anchor='NW'})
    pico.output.draw.layer("layerC", {'!', x=320, y=0, w=150, h=150, anchor='NW'})
    pico.check("tile-grid-06")
end

-- Test 7: layers at different scales
do
    print("3 layers: different scales")
    pico.output.clear()
    pico.output.draw.layer("layerA", {'!', x=0,   y=170, w=120, h=120, anchor='NW'})
    pico.output.draw.layer("layerB", {'!', x=140, y=170, w=200, h=200, anchor='NW'})
    pico.output.draw.layer("layerC", {'!', x=0,   y=310, w=240, h=120, anchor='NW'})
    pico.check("tile-grid-07")
end

pico.init(false)
